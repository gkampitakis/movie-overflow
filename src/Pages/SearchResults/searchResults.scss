.search_results {
  flex-direction: column;
  justify-content: center;
  overflow-y: auto;
}
