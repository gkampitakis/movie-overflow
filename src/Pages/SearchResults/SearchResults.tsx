import React, { useState, useEffect } from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { searchRequest } from '../../Api';
import { SearchOptions, SearchResult } from '../../types';
import ItemList from '../../Components/ItemList';
import ReactPaginate from 'react-paginate';
import './searchResults.scss';

type SearchResultsProps = RouteComponentProps<
  {
    query: string;
  },
  any,
  {
    searchBy: SearchOptions;
  }
>;

export default function SearchResults(props: SearchResultsProps) {
  const [results, setResults] = useState<SearchResult[] | undefined>();
  const [pageCount, setPageCount] = useState(0);
  const [page, setPage] = useState(1);
  const { query } = props.match.params;
  const { searchBy } = props.location.state;

  const handleRequest = (page: number) => {
    searchRequest<{
      results: SearchResult[];
      total_pages: number;
      page: number;
    }>(query, page, searchBy)
      .then((data) => {
        setPage(page);
        setPageCount(data.total_pages);
        setResults(
          data.results.map((res) => ({
            ...res,
            media_type: res.media_type ? res.media_type : (searchBy as any)
          }))
        );
      })
      .catch(console.error);
  };

  useEffect(() => {
    handleRequest(1);
  }, [query, searchBy]);

  return (
    <>
      <section className="main_content">
        <div className="search_results">
          {results?.map((res) => (
            <ItemList item={res} key={res.id} />
          ))}
          {!results?.length && <span>Not results found for {query}</span>}
          {!results && <div>Loading...</div>}
        </div>
        <ReactPaginate
          previousLabel="previous"
          nextLabel="next"
          breakLabel="..."
          pageCount={pageCount}
          marginPagesDisplayed={2}
          pageRangeDisplayed={5}
          onPageChange={({ selected }) => {
            const page = selected + 1;
            setPage(page);
            handleRequest(page);
          }}
          containerClassName={'pagination'}
          activeClassName={'active'}
          forcePage={page - 1}
        />
      </section>
    </>
  );
}
