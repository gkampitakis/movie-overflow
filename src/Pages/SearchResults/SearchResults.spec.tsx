import React from 'react';
import SearchResults from '.';
import { render, act, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FetchMock } from 'jest-fetch-mock';
import {
  searchResponse,
  mockResponseOnce,
  mockDelayedResponseOnce,
  scrollSpy
} from '../../setupTests';

const fetchMock = fetch as FetchMock;
const mockProps = {
  match: {
    params: {
      query: 'mockQuery'
    }
  },
  location: {
    state: undefined
  }
} as any;

describe('SearchResults', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    scrollSpy.mockClear();
    mockResponseOnce(searchResponse.response);
  });

  it('Should show loading', async () => {
    fetchMock.resetMocks();
    mockDelayedResponseOnce({});

    await act(async () => {
      const { container } = render(<SearchResults {...mockProps} />);

      await screen.findByTestId('loader');

      expect(container).toMatchSnapshot();
    });
  });

  it('Should render consistently', async () => {
    const url =
      'https://api.themoviedb.org/3/search/multi?api_key=mockKey&query=mockQuery&page=1';
    await act(async () => {
      const { container } = render(<SearchResults {...mockProps} />);

      await screen.findByAltText(/mockTitle/i);

      expect(fetchMock).toHaveBeenCalledWith(url, {
        method: 'GET'
      });
      expect(container).toMatchSnapshot();
    });
  });

  it('Should show no results message', async () => {
    fetchMock.resetMocks();
    mockResponseOnce(searchResponse.noData);

    await act(async () => {
      const { container } = render(<SearchResults {...mockProps} />);

      await screen.findByAltText(/no results/i);

      expect(container).toMatchSnapshot();
    });
  });

  it('Should change page', async () => {
    mockResponseOnce(searchResponse.response);
    const url1 =
        'https://api.themoviedb.org/3/search/multi?api_key=mockKey&query=mockQuery&page=1',
      url2 =
        'https://api.themoviedb.org/3/search/multi?api_key=mockKey&query=mockQuery&page=2';
    await act(async () => {
      const { container } = render(<SearchResults {...mockProps} />);

      await screen.findByAltText(/mockTitle/i);
      const nextPage = screen.getByText(/next/i);

      userEvent.click(nextPage);

      await screen.findByAltText(/mockTitle/i);

      expect(fetchMock).toHaveBeenCalledWith(url1, {
        method: 'GET'
      });
      expect(fetchMock).toHaveBeenCalledWith(url2, {
        method: 'GET'
      });
      expect(scrollSpy).toBeCalledWith({
        top: 0,
        behavior: 'smooth'
      });
      expect(container).toMatchSnapshot();
    });
  });

  it('Should query by person', async () => {
    const props = {
        ...mockProps,
        location: { state: { searchBy: 'person' } }
      },
      url =
        'https://api.themoviedb.org/3/search/person?api_key=mockKey&query=mockQuery&page=1';

    await act(async () => {
      render(<SearchResults {...props} />);

      await screen.findByAltText(/mockTitle/i);

      expect(fetchMock).toHaveBeenCalledWith(url, {
        method: 'GET'
      });
    });
  });
});
