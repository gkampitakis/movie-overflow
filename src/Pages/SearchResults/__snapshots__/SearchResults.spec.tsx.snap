// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults Should change page 1`] = `
<div>
  <section
    class="main_content"
  >
    <div
      class="search_results"
    >
      <article
        class="item_list"
        data-testid="item-list"
      >
        <img
          alt="mockTitle"
          loading="lazy"
          src="https://image.tmdb.org/t/p/w185/post_path"
        />
        <div
          class="details"
        >
          <div>
            <div
              class="header"
            >
              <h2>
                mockTitle
              </h2>
              <p>
                Movie
              </p>
            </div>
            <p>
              mockDate
            </p>
          </div>
          <p
            class="overview"
          >
            mockOverview
          </p>
        </div>
      </article>
      <div
        class="loader"
        data-testid="loader"
      />
    </div>
    <ul
      class="pagination"
    >
      <li
        class="previous"
      >
        <a
          aria-disabled="false"
          aria-label="Previous page"
          rel="prev"
          role="button"
          tabindex="0"
        >
          previous
        </a>
      </li>
      <li>
        <a
          aria-label="Page 1"
          role="button"
          tabindex="0"
        >
          1
        </a>
      </li>
      <li
        class="active"
      >
        <a
          aria-current="page"
          aria-label="Page 2 is your current page"
          role="button"
          tabindex="0"
        >
          2
        </a>
      </li>
      <li>
        <a
          aria-label="Page 3"
          role="button"
          tabindex="0"
        >
          3
        </a>
      </li>
      <li>
        <a
          aria-label="Page 4"
          role="button"
          tabindex="0"
        >
          4
        </a>
      </li>
      <li>
        <a
          aria-label="Page 5"
          role="button"
          tabindex="0"
        >
          5
        </a>
      </li>
      <li>
        <a
          aria-label="Page 6"
          role="button"
          tabindex="0"
        >
          6
        </a>
      </li>
      <li
        class="break"
      >
        <a
          role="button"
          tabindex="0"
        >
          ...
        </a>
      </li>
      <li>
        <a
          aria-label="Page 9"
          role="button"
          tabindex="0"
        >
          9
        </a>
      </li>
      <li>
        <a
          aria-label="Page 10"
          role="button"
          tabindex="0"
        >
          10
        </a>
      </li>
      <li
        class="next"
      >
        <a
          aria-disabled="false"
          aria-label="Next page"
          rel="next"
          role="button"
          tabindex="0"
        >
          next
        </a>
      </li>
    </ul>
  </section>
</div>
`;

exports[`SearchResults Should render consistently 1`] = `
<div>
  <section
    class="main_content"
  >
    <div
      class="search_results"
    >
      <article
        class="item_list"
        data-testid="item-list"
      >
        <img
          alt="mockTitle"
          loading="lazy"
          src="https://image.tmdb.org/t/p/w185/post_path"
        />
        <div
          class="details"
        >
          <div>
            <div
              class="header"
            >
              <h2>
                mockTitle
              </h2>
              <p>
                Movie
              </p>
            </div>
            <p>
              mockDate
            </p>
          </div>
          <p
            class="overview"
          >
            mockOverview
          </p>
        </div>
      </article>
    </div>
    <ul
      class="pagination"
    >
      <li
        class="previous disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Previous page"
          rel="prev"
          role="button"
          tabindex="0"
        >
          previous
        </a>
      </li>
      <li
        class="active"
      >
        <a
          aria-current="page"
          aria-label="Page 1 is your current page"
          role="button"
          tabindex="0"
        >
          1
        </a>
      </li>
      <li>
        <a
          aria-label="Page 2"
          role="button"
          tabindex="0"
        >
          2
        </a>
      </li>
      <li>
        <a
          aria-label="Page 3"
          role="button"
          tabindex="0"
        >
          3
        </a>
      </li>
      <li>
        <a
          aria-label="Page 4"
          role="button"
          tabindex="0"
        >
          4
        </a>
      </li>
      <li>
        <a
          aria-label="Page 5"
          role="button"
          tabindex="0"
        >
          5
        </a>
      </li>
      <li>
        <a
          aria-label="Page 6"
          role="button"
          tabindex="0"
        >
          6
        </a>
      </li>
      <li
        class="break"
      >
        <a
          role="button"
          tabindex="0"
        >
          ...
        </a>
      </li>
      <li>
        <a
          aria-label="Page 9"
          role="button"
          tabindex="0"
        >
          9
        </a>
      </li>
      <li>
        <a
          aria-label="Page 10"
          role="button"
          tabindex="0"
        >
          10
        </a>
      </li>
      <li
        class="next"
      >
        <a
          aria-disabled="false"
          aria-label="Next page"
          rel="next"
          role="button"
          tabindex="0"
        >
          next
        </a>
      </li>
    </ul>
  </section>
</div>
`;

exports[`SearchResults Should show loading 1`] = `
<div>
  <section
    class="main_content"
  >
    <div
      class="search_results"
    >
      <div
        class="loader"
        data-testid="loader"
      />
    </div>
  </section>
</div>
`;

exports[`SearchResults Should show no results message 1`] = `
<div>
  <section
    class="main_content"
  >
    <div
      class="search_results"
    >
      <article
        class="image_container no_results"
      >
        <img
          alt="No results returned"
          src="noData.svg"
        />
        <p>
          No results found for '
          mockQuery
          '
        </p>
        <span>
          ðŸ˜­
        </span>
      </article>
    </div>
  </section>
</div>
`;
